File: task/src/main/java/dojo/Board.java
package dojo;

import java.util.ArrayList;
import java.util.List;

import static java.util.stream.Collectors.toList;

public class Board {
    private final List<List<Cell>> grid;

    public Board(List<List<Cell>> grid) {
        this.grid = grid;
    }

    public Board(Board board) {
        grid = new ArrayList<>(board.size());
        for (int row = 0; row < board.size(); row++) {
            grid.add(new ArrayList<>(board.grid.get(row)));
        }
    }

    public Board set(int row, int column, Cell value) {
        List<List<Cell>> gridCopy = grid.stream().map(ArrayList::new).collect(toList());
        gridCopy.get(row).set(column, value);
        return new Board(gridCopy);
    }

    public Cell get(int row, int column) {
        return grid.get(row).get(column);
    }

    public int size() {
        return grid.size();
    }
}

File: 12_Noughts_and_Crosses_not_quite/src/main/kotlin/dojo/Board.kt
package dojo

import java.util.*

class Board(private val grid: List<List<Cell>>) {

    constructor(board: Board) : this(board.grid)

    operator fun set(row: Int, column: Int, value: Cell): Board {
        val gridCopy = grid.map { ArrayList(it) }
        gridCopy[row][column] = value
        return Board(gridCopy)
    }

    operator fun get(row: Int, column: Int) = grid[row][column]

    fun size() = grid.size

    fun isFull(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun all(coordinates: List<Pair<Int, Int>>, cell: Cell): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
